src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -442,23 +453,39 @@ public void run() throws InterruptedException {
                 // operations. This ensures the shard engine is started and disables
                 // garbage collection (not the JVM's GC!) of tombstone deletes
                 transportService.submitRequest(request.targetNode(), RecoveryTarget.Actions.PREPARE_TRANSLOG,
-                        new RecoveryPrepareForTranslogOperationsRequest(request.recoveryId(), request.shardId(), shard.translog().estimatedNumberOfOperations()),
+                        new RecoveryPrepareForTranslogOperationsRequest(request.recoveryId(), request.shardId(), translogView.estimatedTotalOperations()),
                         TransportRequestOptions.options().withTimeout(recoverySettings.internalActionTimeout()), EmptyTransportResponseHandler.INSTANCE_SAME).txGet();
             }
         });
 
         stopWatch.stop();
-        response.startTime = stopWatch.totalTime().millis();
-        logger.trace("{} recovery [phase2] to {}: start took [{}]",
+
+        response.startTime = stopWatch.totalTime().millis() - startEngineStart;
+        logger.trace("{} recovery [phase1] to {}: remote engine start took [{}]",
                 request.shardId(), request.targetNode(), stopWatch.totalTime());
+    }
 
+    /**
+     * Perform phase2 of the recovery process
+     *
+     * Phase2 takes a snapshot of the current translog *without* acquiring the
+     * write lock (however, the translog snapshot is a point-in-time view of
+     * the translog). It then sends each translog operation to the target node
+     * so it can be replayed into the new shard.
+     *
+     */
+    public void phase2(Translog.Snapshot snapshot) throws ElasticsearchException {
+        if (shard.state() == IndexShardState.CLOSED) {
+            throw new IndexShardClosedException(request.shardId());
+        }
+        cancellableThreads.checkForCancel();
 
+        StopWatch stopWatch = new StopWatch().start();
         logger.trace("{} recovery [phase2] to {}: updating current mapping to master", request.shardId(), request.targetNode());
         // Ensure that the mappings are synced with the master node
         updateMappingOnMaster();
```

Comment:
Would be fantastic to have a stopwatch capture how long updating the mappings took, since you moved it to before this call it will now include this, but I would love to separate them into two different timings.